stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,553 @@
+				},
+				Required: semantic.LabelSet{"left", "right"},
+				Return:   semantic.NewEmptyObjectPolyType(),
+			}),
+		},
+		Required: semantic.LabelSet{"left", "right", "fn"},
+		Return:   flux.TableObjectType,
+	}
+	flux.RegisterPackageValue("internal/promql", "join", flux.FunctionValue("join", createJoinOpSpec, signature))
+	flux.RegisterOpSpec(joinKind, newJoinOp)
+	plan.RegisterProcedureSpec(joinKind, newMergeJoinProcedure, joinKind)
+	execute.RegisterTransformation(joinKind, createMergeJoinTransformation)
+}
+
+type JoinOpSpec struct {
+	Left  flux.OperationID             `json:"left"`
+	Right flux.OperationID             `json:"right"`
+	Fn    interpreter.ResolvedFunction `json:"fn"`
+
+	l, r *flux.TableObject
+}
+
+func (s *JoinOpSpec) IDer(ider flux.IDer) {
+	s.Left = ider.ID(s.l)
+	s.Right = ider.ID(s.r)
+}
+
+func createJoinOpSpec(args flux.Arguments, p *flux.Administration) (flux.OperationSpec, error) {
+	l, err := args.GetRequiredObject("left")
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(josh): The type system should ensure that this
+	// assertion is redundant. Unfortunately it does not.
+	// Remove this check when type inference is fixed.
+	//
+	left, ok := l.(*flux.TableObject)
+	if !ok {
+		return nil, errors.New(codes.Invalid, "argument 'left' must be a table stream")
+	}
+	p.AddParent(left)
+
+	r, err := args.GetRequiredObject("right")
+	if err != nil {
+		return nil, err
+	}
+
+	// Same comment as above. The type system should ensure
+	// that the folowing cast never panics.
+	//
+	right, ok := r.(*flux.TableObject)
+	if !ok {
+		return nil, errors.New(codes.Invalid, "argument 'right' must be a table stream")
+	}
+	p.AddParent(right)
+
+	f, err := args.GetRequiredFunction("fn")
+	if err != nil {
+		return nil, err
+	}
+
+	fn, err := interpreter.ResolveFunction(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return &JoinOpSpec{
+		Fn: fn,
+		l:  left,
+		r:  right,
+	}, nil
+}
+
+func newJoinOp() flux.OperationSpec {
+	return new(JoinOpSpec)
+}
+
+func (s *JoinOpSpec) Kind() flux.OperationKind {
+	return joinKind
+}
+
+type MergeJoinProcedureSpec struct {
+	plan.DefaultCost
+
+	Fn interpreter.ResolvedFunction `json:"fn"`
+}
+
+func newMergeJoinProcedure(spec flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	s, ok := spec.(*JoinOpSpec)
+	if !ok {
+		return nil, errors.Newf(codes.Invalid, "invalid spec type %T", spec)
+	}
+	return &MergeJoinProcedureSpec{Fn: s.Fn}, nil
+}
+
+func (s *MergeJoinProcedureSpec) Kind() plan.ProcedureKind {
+	return joinKind
+}
+func (s *MergeJoinProcedureSpec) Copy() plan.ProcedureSpec {
+	return &MergeJoinProcedureSpec{Fn: s.Fn.Copy()}
+}
+
+func createMergeJoinTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	s, ok := spec.(*MergeJoinProcedureSpec)
+	if !ok {
+		return nil, nil, errors.Newf(codes.Invalid, "invalid spec type %T", spec)
```

Comment:
This should be an internal error. There's nothing the user can do about this and it indicates a logic error.